### PR TITLE
[CGP-424] Add emulator events

### PR DESCRIPTION
**Pull request by j-mueller**
* Emulator keeps a log of blockchain-related events such as transaction
  validation

### CODE REVIEW

File: wallet-api/src/Wallet/Emulator/Log.hs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    ) where
+
+import           Data.Aeson        (FromJSON, ToJSON)
+import           GHC.Generics      (Generic)
+
+import qualified Wallet.UTXO.Index as UTXO
+import qualified Wallet.UTXO.Types as UTXO
+
+-- | Events produced by the mockchain
+data EmulatorEvent =
```
[the remark this answers is not in the document]
Comment (michaelpj) [reply wallet-api/src/Wallet/Emulator/Log.hs]: I think this also gives us an additional nice layering: we have a trace of actions, which turns into a trace of state changes, which can be applied to actually compute an updated state.

[the remark this answers is not in the document]
Comment (michaelpj) [reply wallet-api/src/Wallet/Emulator/Log.hs]: That's a state change, right? It moves the transaction from the pending pool to the chain.

[the remark this answers is not in the document]
Comment (j-mueller) [reply wallet-api/src/Wallet/Emulator/Log.hs]: OK, I'll make the change and we'll see what it looks like.

[the remark this answers is not in the document]
Comment (j-mueller) [reply wallet-api/src/Wallet/Emulator/Log.hs]: Check out [Wallet.Emulator.TypesNew](https://github.com/input-output-hk/plutus/pull/320/commits/feee851f6c3475cf99ad9f7a21bcd6b47e47b08f#diff-3b05b949145accd601e8dab80cd75e4e), it's not exactly what you said but it does achieve some breaking up of the emulator into parts

[the remark this answers is not in the document]
Comment (michaelpj) [reply wallet-api/src/Wallet/Emulator/Log.hs]: Btw if this doesn't seem productive right now feel free to drop it and we can revisit it later.

[the remark this answers is not in the document]
Comment (michaelpj) [reply wallet-api/src/Wallet/Emulator/Log.hs]: Looks promising! So we should be able to interpret the various events as modifying various bits of the state, and then the overall state just glues them all together.

[the remark this answers is not in the document]
Comment (j-mueller) [reply wallet-api/src/Wallet/Emulator/Log.hs]: > Btw if this doesn't seem productive right now feel free to drop it and we can revisit it later.

It shouldn't take too much longer now so I'll just finish it

[the remark this answers is not in the document]
Comment (j-mueller) [reply wallet-api/src/Wallet/Emulator/Log.hs]: Actually it is going to take much longer so I think we should merge this PR as it is (after I have rebased it) and then I'll get back to the event based thing later. I've put it on a different branch, https://github.com/input-output-hk/plutus/tree/j-mueller/emulator-types-new

[the remark this answers is not in the document]
Comment (michaelpj) [reply wallet-api/src/Wallet/Emulator/Log.hs]: Fair enough - it was just an idea! I've been thinking about it a bit too - we also have a bit of a state machine going on here.

**Pull request merged**